# Relationships Between Tables

## **Getting Started**

This week we will be learning how to how to pull data from multiple tables.

**Before beginning:** It is recommended that you open/download the _guitar store ERD_ that is available in Github so you can review it during the assignment.

## **Joins**

Joins are how we relate tables during query execution to obtain a specific result. Joins are different from relationships in the sense that relationships exist for the purpose of maintaining data integrity between tables. The relationship that exist in the _guitar store_ database between _Categories_ and _Products_ is in place to ensure that there are no products that don't have a corresponding record in the _Categories_ table. To return related data between these tables we can use the JOIN statement in our SQL.

During this exercise, we will be using the _Products_ and _Categories_ tables. Run the below code cells (2) to see the structure and data in each table.

In [1]:
--Select all records form gs_products
SELECT *
FROM gs_products

: Msg 208, Level 16, State 1, Line 2
Invalid object name 'gs_products'.

Total execution time: 00:00:00.261

In [4]:
--Select all records form gs_categories
SELECT *
FROM gs_categories

: Msg 208, Level 16, State 1, Line 2
Invalid object name 'gs_categories'.

Total execution time: 00:00:00.273

We can tell from the results of our queries and the _guitar shop_ ERD that instead of storing the category name inside of the product table, we store a category ID and use a relationship/join to look for that category ID in the categories table. There are two main reasons we do this:

- It takes more space to store the word "Basses" than the integer 2. Please see the storage sizes below. This is a massive difference. Instead of storing the word "Basses" everytime that we need it, we can store it one time in the category table and use the ID it was assigned to pull it into our result.
    - 2 in binary: 10
    - "Basses" in binary: 01100010 01100001 01110011 01110011 01100101 01110011
    - So, it takes 24 times as much storage to store the word "Basses" than it does to store the number 2.
- The other reason is that it helps us preserve data integrity. Due to referencial integrity, anything we put in Products.CategoryID must have a matching value in Categories.CategoryID. This prevents erraneous values from sneaking in like "Bassses", "Bases", etc...

### **Inner Join**

Using the JOIN statement in SQL to pull data from multiple tables results in an _inner join._ An inner join will only return records/rows of the selected fields from tables where the joined fields are equal. For example, _Categories_ and _Products_ can be joined on _CategoryID_ since that field is a FK in the _Products_ table. In the below code cell we are selecting all records from both tables where the join fields are equal.

In [ ]:
--Return all records from products where there is a match in categories
SELECT p.*, c.*
FROM gs_products as p 
JOIN gs_categories as c ON p.category_id = c.category_id

When running the above code cell we can see that the result returns all fields from both tables in the order that we entered them into the select statatement. The below code cell contains a query that will just pull in a few fields from each table to evaluate our join.

In [ ]:
--Return specific columns from products where there is a match in categories
SELECT p.product_id, p.product_code, p.category_id AS products_category_id, c.category_id AS categories_category_id, c.category_name
FROM gs_products AS p
JOIN gs_categories AS c ON p.category_id = c.category_id

The above result shows us that we are brining in data from the categories table where Categories.CategoryID matches Products.CategoryID. We know from our earlier example that Category ID 2 is "Basses", so "Basses" is returned as the CategoryName for any products with a 2 in the categoryID. We can also use a WHERE clause to limit the return of data in the products table when joining on the categories table.

In [ ]:
--Return all records from products where the product is a Base
SELECT *
FROM gs_products AS p
JOIN gs_categories AS c ON p.category_id = c.category_id
WHERE c.category_name = 'Basses'

The above code cell returns all products where the category name is "Basses".  Even though we aren't brining in any of the data from the categories table into our select statement, we can still refernce it in our from clause through a join and filter it with criteria in our where clause.

✏️ **Practice**

Edit the below code cell so that the query ruturns all data from the products table that are in the _guitar_ category by using a join.

In [40]:
SELECT p.*
FROM gs_products AS p
JOIN gs_categories AS c
    ON p.category_id = c.category_id
WHERE c.category_name = 'Guitars';

(6 rows affected)

Total execution time: 00:00:00.294

product_id category_id product_code product_name product_description list_price discount_percent date_added 1 1 strat Fender Stratocaster The Fender Stratocaster is the electric guitar design that changed the world. New features include a tinted neck, parchment pickguard and control knobs, and a '70s-style logo. Includes select alder body, 21-fret maple neck with your choice of a rosewood or maple fretboard, 3 single-coil pickups, vintage-style tremolo, and die-cast tuning keys. This guitar features a thicker bridge block for increased sustain and a more stable point of contact with the strings. At this low price, why play anything but the real thing?\r\n\r\nFeatures:\r\n\r\n* New features:\r\n* Thicker bridge block\r\n* 3-ply parchment pick guard\r\n* Tinted neck 699.00 30.00 2015-10-30 09:32:40.000 2 1 les_paul Gibson Les Paul This Les Paul guitar offers a carved top and humbucking pickups. It has a simple yet elegant design. Cutting-yet-rich tone?the hallmark of the Les Paul?pours out of the 490R and 498T Alnico II magnet humbucker pickups, which are mounted on a carved maple top with a mahogany back. The faded finish models are equipped with BurstBucker Pro pickups and a mahogany top. This guitar includes a Gibson hardshell case (Faded and satin finish models come with a gig bag) and a limited lifetime warranty.\r\n\r\nFeatures:\r\n\r\n* Carved maple top and mahogany back (Mahogany top on faded finish models)\r\n* Mahogany neck, '59 Rounded Les Paul\r\n* Rosewood fingerboard (Ebony on Alpine white)\r\n* Tune-O-Matic bridge with stopbar\r\n* Chrome or gold hardware\r\n* 490R and 498T Alnico 2 magnet humbucker pickups (BurstBucker Pro on faded finish models)\r\n* 2 volume and 2 tone knobs, 3-way switch 1199.00 30.00 2015-12-05 16:33:13.000 3 1 sg Gibson SG This Gibson SG electric guitar takes the best of the '62 original and adds the longer and sturdier neck joint of the late '60s models. All the classic features you'd expect from a historic guitar. Hot humbuckers go from rich, sweet lightning to warm, tingling waves of sustain. A silky-fast rosewood fretboard plays like a dream. The original-style beveled mahogany body looks like a million bucks. Plus, Tune-O-Matic bridge and chrome hardware. Limited lifetime warranty. Includes hardshell case.\r\n\r\nFeatures:\r\n\r\n* Double-cutaway beveled mahogany body\r\n* Set mahogany neck with rounded '50s profile\r\n* Bound rosewood fingerboard with trapezoid inlays\r\n* Tune-O-Matic bridge with stopbar tailpiece\r\n* Chrome hardware\r\n* 490R humbucker in the neck position\r\n* 498T humbucker in the bridge position\r\n* 2 volume knobs, 2 tone knobs, 3-way switch\r\n* 24-3/4" scale 2517.00 52.00 2016-02-04 11:04:31.000 4 1 fg700s Yamaha FG700S The Yamaha FG700S solid top acoustic guitar has the ultimate combo for projection and pure tone. The expertly braced spruce top speaks clearly atop the rosewood body. It has a rosewood fingerboard, rosewood bridge, die-cast tuners, body and neck binding, and a tortoise pickguard.\r\n\r\nFeatures:\r\n\r\n* Solid Sitka spruce top\r\n* Rosewood back and sides\r\n* Rosewood fingerboard\r\n* Rosewood bridge\r\n* White/black body and neck binding\r\n* Die-cast tuners\r\n* Tortoise pickguard\r\n* Limited lifetime warranty 489.99 38.00 2016-06-01 11:12:59.000 5 1 washburn Washburn D10S The Washburn D10S acoustic guitar is superbly crafted with a solid spruce top and mahogany back and sides for exceptional tone. A mahogany neck and rosewood fingerboard make fretwork a breeze, while chrome Grover-style machines keep you perfectly tuned. The Washburn D10S comes with a limited lifetime warranty.\r\n\r\nFeatures:\r\n\r\n * Spruce top\r\n * Mahogany back, sides\r\n * Mahogany neck Rosewood fingerboard\r\n * Chrome Grover-style machines 299.00 0.00 2016-07-30 13:58:35.000 6 1 rodriguez Rodriguez Caballero 11 Featuring a carefully chosen, solid Canadian cedar top and laminated bubinga back and sides, the Caballero 11 classical guitar is a beauty to behold and 

**Expected Result**



### **Left/Right Joins**

While inner joins return all records from both tables where there is a match in the joining fields, a left or right join will return all records from the specified table, and only matching records from the joined table. Using the next few code cells, we will build a query that shows all customers who have never placed an order using a left join. Before we do so, run the code cells (2) below to view the data and structure of the customers and orders tables.

In [16]:
--Return all records from customers
SELECT *
FROM gs_customers

(485 rows affected)

Total execution time: 00:00:01.014

customer_id email_address customer_password first_name last_name shipping_address_id billing_address_id 1 allan.sherwood@yahoo.com c44321e51ec184a2f739318639cec426de774451 Allan Sherwood 1 2 2 barryz@gmail.com d9e03c0b34c57d034edda004ec8bae5d53667e36 Barry Zimmer 3 3 3 christineb@solarone.com 13ef4f968693bda97a898ece497da087b182808e Christine Brown 4 4 4 david.goldstein@hotmail.com 2a367cbb171d78d293f40fd7d1defb31e3fb1728 David Goldstein 5 6 5 erinv@gmail.com 2e203dd22e39e3a8930e7641fe074fec2b18b102 Erin Valentino 7 7 6 frankwilson@sbcglobal.net b13773cfee62f832cacb618b257feec972f30b13 Frank Lee Wilson 8 8 7 gary_hernandez@yahoo.com e931eea39d638c0324c0065c40e2c0acc91ceca9 Gary Hernandez 9 10 8 heatheresway@mac.com 1867b31afdfbb6814133aa545f67305bc2f211b6 Heather Esway 11 12 9 jbutt@gmail.com 28592e5b54a8babc3cb6ad0a1c9094a2621c8ac3 James Butt 13 13 10 josephine_darakjy@darakjy.org 815d965d07c98821d8ca725243bd09def4e33f24 Josephine Darakjy 14 14 11 art@venere.org 39f7cfd0b51970b6bcd55a00da8e672b7153a183 Art Venere 15 16 12 lpaprocki@hotmail.com af27e4c5b48549d6dee6199e56834af1404f6169 Lenna Paprocki 17 17 13 donette.foller@cox.net 356584acaa7164280c97b147f317ce0d76d5993c Donette Foller 18 18 14 simona@morasca.com 81133e972ea9b42831ef1fe7a846493526bfd3c3 Simona Morasca 19 19 15 mitsue_tollner@yahoo.com 460420ee0d4f211f6be0bf7e7f352cf135cac579 Mitsue Tollner 20 21 16 leota@hotmail.com ceb92dd2b07d22fa39b9188306b9b04a78d80fe8 Leota Dilliard 22 22 17 sage_wieser@cox.net 0c317d5469cda724d8b81ccd989179c99c3f2563 Sage Wieser 23 23 18 kris@gmail.com d79af41c43ea2a02afbbd66af85a70bed5b5757f Kris Marrier 24 24 19 minna_amigon@yahoo.com c73630df7813e0cc99e02c639bdce1a954257230 Minna Amigon 25 26 20 amaclead@gmail.com 295474f02de5da5862b837ca53de5e850f37f4c5 Abel Maclead 27 27 21 kiley.caldarera@aol.com eb3cc1df12acbdd0f6f39c56891b12f588cfb18a Kiley Caldarera 28 28 22 gruta@cox.net e26bf3250ec4d8c764a06235f9a8aaac0354050c Graciela Ruta 29 29 23 calbares@gmail.com effd5ca2705a8a11fe6b79eff386b45dd2ad8d5c Cammy Albares 30 30 24 mattie@aol.com 674de46b21da5e95aad11c3e24e7ddae7dd063f7 Mattie Poquette 31 31 25 meaghan@hotmail.com de6aac016c911c349eeda6c7ce8827c10cd6003a Meaghan Garufi 32 32 26 gladys.rim@rim.org db4f92cbe6e82c70a754045a98eda9d2906fe960 Gladys Rim 33 33 27 yuki_whobrey@aol.com df9b0a27fddfb22ebef677ff17bb2630b364b883 Yuki Whobrey 34 34 28 fletcher.flosi@yahoo.com 0a73bcca3abb8abe8c7af206e7a173f80ff37471 Fletcher Flosi 35 35 29 bette_nicka@cox.net aa18a90d953be48bb88ba61b79bccb694dacac7a Bette Nicka 36 37 30 vinouye@aol.com 8e2b3b9115757100057cf291b2d31d8fb5a6ff57 Veronika Inouye 38 38 31 willard@hotmail.com 1b437e1d7a706a108271636d66487b065b359579 Willard Kolmetz 39 39 32 mroyster@royster.com f7046c37b0ad80622b956464ee41e590e4b833dd Maryann Royster 40 41 33 alisha@slusarski.com ca4db1a43bcab5dc00c6d2afddc5f85f14a27a1b Alisha Slusarski 42 43 34 allene_iturbide@cox.net e7c8ddb5894ecbf48b9285d3d480fa153377d106 Allene Iturbide 44 44 35 chanel.caudy@caudy.org c4940cd601f8ec0bb2dd4e40dcaead38496fbf56 Chanel Caudy 45 45 36 ezekiel@chui.com 28bb43c8489d35ae5e0410a3e7f2920184f86d79 Ezekiel Chui 46 46 37 wkusko@yahoo.com 3769a748f05734c0692cd7dcb1718b39dc0fdc52 Willow Kusko 47 47 38 bfigeroa@aol.com b016c40c80f68317b242d07469d65c6ba07310dd Bernardo Figeroa 48 48 39 ammie@corrio.com 88f982f0ad4f45293d4adb30294fe7407cf99c01 Ammie Corrio 49 49 40 francine_vocelka@vocelka.com 94e3fe0c1a2a02d28c64da77b339f14308f2ee29 Francine Vocelka 50 51 41 ernie_stenseth@aol.com 4944f2cfff1dc49827539f415a390c56d6185fbd Ernie Stenseth 52 52 42 albina@glick.com afc12cca2e0eb72a0d8ca7849ac2b8687fef3e63 Albina Glick 53 53 43 asergi@gmail.com 8b06eecc3b3ac8aef4199354845a340b5776b0fc Alishia Sergi 54 54 44 solange@shinko.com b067b71aff361ba9b6d1eadd27e52eebdcde26ca Solange Shinko 55 55 45 jose@yahoo.com 866b8c2845010932a0e954c8596d52791c9dd83d Jose Stockham 56 56 46 rozella.ostrosky@ostrosky.com 0c6e3b007e144cd464786b81a5e3b51d1eb9b1be Rozella Ostrosky 57 57 47 

In [17]:
--Return all records from orders
SELECT *
FROM gs_orders

(41 rows affected)

Total execution time: 00:00:00.312

order_id customer_id order_date ship_amount tax_amount ship_date ship_address_id card_type card_number card_expires billing_address_id 1 1 2016-03-28 09:40:28.000 5.00 58.75 2016-03-31 09:41:11.000 1 Visa 4111111111111111 04/2018 2 2 2 2016-03-28 11:23:20.000 5.00 21.27 2016-03-31 11:24:03.000 3 Visa 4012888888881881 08/2020 3 3 1 2016-03-29 09:44:58.000 10.00 102.29 2016-04-01 09:45:41.000 1 Visa 4111111111111111 06/2020 2 4 3 2016-03-30 15:22:31.000 10.00 117.50 2016-04-02 15:23:14.000 4 American Express 3782822463100005 02/2017 4 5 4 2016-03-31 05:43:11.000 5.00 20.93 2016-04-03 05:43:54.000 5 Visa 4111111111111111 09/2019 6 6 5 2016-03-31 18:37:22.000 5.00 20.93 2016-04-03 18:38:05.000 7 Discover 6011111111111117 04/2020 7 7 6 2016-04-01 23:11:12.000 15.00 107.80 2016-04-04 23:11:55.000 8 MasterCard 5555555555554444 12/2018 8 8 7 2016-04-02 11:26:38.000 5.00 47.60 2016-04-05 11:27:21.000 9 Visa 4012888888881881 04/2017 10 9 4 2016-04-03 12:22:31.000 15.00 102.75 2016-04-06 12:23:14.000 5 Visa 4111111111111111 01/2020 6 10 8 2016-04-03 14:59:20.000 5.00 26.25 2016-04-06 15:00:03.000 11 Visa 4111111111111111 08/2019 12 11 9 2016-04-04 06:24:44.000 5.00 34.25 2016-04-07 06:25:27.000 13 Visa 4012888888881881 08/2019 13 12 10 2016-04-04 08:15:12.000 5.00 84.57 2016-04-07 08:15:55.000 14 Visa 4111111111111111 03/2017 14 13 11 2016-04-04 11:20:31.000 5.00 47.60 2016-04-07 11:21:14.000 15 Visa 4111111111111111 02/2020 16 14 12 2016-04-05 09:24:53.000 10.00 117.50 2016-04-08 09:25:36.000 17 Visa 4111111111111111 11/2018 17 15 13 2016-04-05 14:52:17.000 5.00 39.20 2016-04-08 14:53:00.000 18 American Express 3782822463100005 02/2018 18 16 14 2016-04-06 07:53:42.000 10.00 51.97 2016-04-09 07:54:25.000 19 Visa 4111111111111111 01/2019 19 17 15 2016-04-06 17:24:28.000 5.00 34.25 2016-04-09 17:25:11.000 20 Visa 4111111111111111 07/2020 21 18 16 2016-04-06 18:41:53.000 5.00 34.25 2016-04-09 18:42:36.000 22 MasterCard 5555555555554444 12/2017 22 19 17 2016-04-08 12:21:31.000 10.00 117.50 2016-04-11 12:22:14.000 23 Visa 4012888888881881 12/2017 23 20 18 2016-04-10 09:33:23.000 5.00 47.60 2016-04-13 09:34:06.000 24 Visa 4111111111111111 05/2017 24 21 19 2016-04-11 08:21:32.000 10.00 39.20 2016-04-14 08:22:15.000 25 Discover 6011111111111117 02/2019 26 22 8 2016-04-12 12:26:52.000 5.00 84.57 2016-04-15 12:27:35.000 11 Visa 4111111111111111 03/2020 12 23 20 2016-04-14 07:59:31.000 5.00 34.25 2016-04-17 08:00:14.000 27 Visa 4012888888881881 03/2019 27 24 21 2016-04-17 17:40:22.000 5.00 34.25 2016-04-20 17:41:05.000 28 Visa 4111111111111111 04/2018 28 25 22 2016-04-20 08:23:32.000 10.00 117.50 2016-04-23 08:24:15.000 29 Visa 4111111111111111 09/2018 29 26 23 2016-04-20 08:14:45.000 5.00 0.00 2016-04-23 08:15:28.000 30 American Express 3782822463100005 08/2017 30 27 24 2016-04-20 09:17:52.000 5.00 84.57 2016-04-23 09:18:35.000 31 Visa 4111111111111111 02/2018 31 28 25 2016-04-21 17:52:24.000 5.00 34.30 2016-04-24 17:53:07.000 32 Visa 4111111111111111 08/2019 32 29 4 2016-04-25 23:36:41.000 25.00 196.00 2016-04-28 23:37:24.000 5 Visa 4012888888881881 03/2017 6 30 26 2016-04-27 16:21:31.000 5.00 26.25 2016-04-30 16:22:14.000 33 Visa 4111111111111111 02/2018 33 31 27 2016-04-29 06:47:14.000 10.00 118.82 2016-05-02 06:47:57.000 34 Visa 4111111111111111 01/2018 34 32 18 2016-05-01 01:23:23.000 5.00 84.57 NULL 24 Discover 6011111111111117 02/2018 24 33 28 2016-05-01 09:11:51.000 10.00 41.86 2016-05-04 09:12:34.000 35 American Express 3782822463100005 04/2017 35 34 29 2016-05-02 11:36:12.000 5.00 58.75 2016-05-05 11:36:55.000 36 Visa 4111111111111111 06/2017 37 35 30 2016-05-04 03:52:23.000 5.00 39.20 2016-05-07 03:53:06.000 38 Visa 4111111111111111 09/2018 38 36 31 2016-05-04 12:31:33.000 5.00 21.27 2016-05-07 12:32:16.000 39 Visa 4012888888881881 11/2018 39 37 32 2016-05-06 14:15:21.000 5.00 84.57 2016-05-09 14:16:04.000 40 MasterCard 5555555555554444 02/2017 41 38 33 2016-05-08 11:41:24.000 10.00 117.50 NULL 42 Visa 4111111111111111 04/2018 43 39 29 

We can see from the above queries that there are 486 customers and 41 orders. Let's start working on our "Customers without orders" query by creating a query that joins the two tables in customerID to see which customers have placed an order.

In [18]:
--Return all customers who appear in the orders table
SELECT c.*, o.*
FROM gs_customers as c
JOIN gs_orders as o ON c.customer_id = o.customer_id

(41 rows affected)

Total execution time: 00:00:00.403

customer_id email_address customer_password first_name last_name shipping_address_id billing_address_id order_id customer_id order_date ship_amount tax_amount ship_date ship_address_id card_type card_number card_expires billing_address_id 1 allan.sherwood@yahoo.com c44321e51ec184a2f739318639cec426de774451 Allan Sherwood 1 2 1 1 2016-03-28 09:40:28.000 5.00 58.75 2016-03-31 09:41:11.000 1 Visa 4111111111111111 04/2018 2 2 barryz@gmail.com d9e03c0b34c57d034edda004ec8bae5d53667e36 Barry Zimmer 3 3 2 2 2016-03-28 11:23:20.000 5.00 21.27 2016-03-31 11:24:03.000 3 Visa 4012888888881881 08/2020 3 1 allan.sherwood@yahoo.com c44321e51ec184a2f739318639cec426de774451 Allan Sherwood 1 2 3 1 2016-03-29 09:44:58.000 10.00 102.29 2016-04-01 09:45:41.000 1 Visa 4111111111111111 06/2020 2 3 christineb@solarone.com 13ef4f968693bda97a898ece497da087b182808e Christine Brown 4 4 4 3 2016-03-30 15:22:31.000 10.00 117.50 2016-04-02 15:23:14.000 4 American Express 3782822463100005 02/2017 4 4 david.goldstein@hotmail.com 2a367cbb171d78d293f40fd7d1defb31e3fb1728 David Goldstein 5 6 5 4 2016-03-31 05:43:11.000 5.00 20.93 2016-04-03 05:43:54.000 5 Visa 4111111111111111 09/2019 6 5 erinv@gmail.com 2e203dd22e39e3a8930e7641fe074fec2b18b102 Erin Valentino 7 7 6 5 2016-03-31 18:37:22.000 5.00 20.93 2016-04-03 18:38:05.000 7 Discover 6011111111111117 04/2020 7 6 frankwilson@sbcglobal.net b13773cfee62f832cacb618b257feec972f30b13 Frank Lee Wilson 8 8 7 6 2016-04-01 23:11:12.000 15.00 107.80 2016-04-04 23:11:55.000 8 MasterCard 5555555555554444 12/2018 8 7 gary_hernandez@yahoo.com e931eea39d638c0324c0065c40e2c0acc91ceca9 Gary Hernandez 9 10 8 7 2016-04-02 11:26:38.000 5.00 47.60 2016-04-05 11:27:21.000 9 Visa 4012888888881881 04/2017 10 4 david.goldstein@hotmail.com 2a367cbb171d78d293f40fd7d1defb31e3fb1728 David Goldstein 5 6 9 4 2016-04-03 12:22:31.000 15.00 102.75 2016-04-06 12:23:14.000 5 Visa 4111111111111111 01/2020 6 8 heatheresway@mac.com 1867b31afdfbb6814133aa545f67305bc2f211b6 Heather Esway 11 12 10 8 2016-04-03 14:59:20.000 5.00 26.25 2016-04-06 15:00:03.000 11 Visa 4111111111111111 08/2019 12 9 jbutt@gmail.com 28592e5b54a8babc3cb6ad0a1c9094a2621c8ac3 James Butt 13 13 11 9 2016-04-04 06:24:44.000 5.00 34.25 2016-04-07 06:25:27.000 13 Visa 4012888888881881 08/2019 13 10 josephine_darakjy@darakjy.org 815d965d07c98821d8ca725243bd09def4e33f24 Josephine Darakjy 14 14 12 10 2016-04-04 08:15:12.000 5.00 84.57 2016-04-07 08:15:55.000 14 Visa 4111111111111111 03/2017 14 11 art@venere.org 39f7cfd0b51970b6bcd55a00da8e672b7153a183 Art Venere 15 16 13 11 2016-04-04 11:20:31.000 5.00 47.60 2016-04-07 11:21:14.000 15 Visa 4111111111111111 02/2020 16 12 lpaprocki@hotmail.com af27e4c5b48549d6dee6199e56834af1404f6169 Lenna Paprocki 17 17 14 12 2016-04-05 09:24:53.000 10.00 117.50 2016-04-08 09:25:36.000 17 Visa 4111111111111111 11/2018 17 13 donette.foller@cox.net 356584acaa7164280c97b147f317ce0d76d5993c Donette Foller 18 18 15 13 2016-04-05 14:52:17.000 5.00 39.20 2016-04-08 14:53:00.000 18 American Express 3782822463100005 02/2018 18 14 simona@morasca.com 81133e972ea9b42831ef1fe7a846493526bfd3c3 Simona Morasca 19 19 16 14 2016-04-06 07:53:42.000 10.00 51.97 2016-04-09 07:54:25.000 19 Visa 4111111111111111 01/2019 19 15 mitsue_tollner@yahoo.com 460420ee0d4f211f6be0bf7e7f352cf135cac579 Mitsue Tollner 20 21 17 15 2016-04-06 17:24:28.000 5.00 34.25 2016-04-09 17:25:11.000 20 Visa 4111111111111111 07/2020 21 16 leota@hotmail.com ceb92dd2b07d22fa39b9188306b9b04a78d80fe8 Leota Dilliard 22 22 18 16 2016-04-06 18:41:53.000 5.00 34.25 2016-04-09 18:42:36.000 22 MasterCard 5555555555554444 12/2017 22 17 sage_wieser@cox.net 0c317d5469cda724d8b81ccd989179c99c3f2563 Sage Wieser 23 23 19 17 2016-04-08 12:21:31.000 10.00 117.50 2016-04-11 12:22:14.000 23 Visa 4012888888881881 12/2017 23 18 kris@gmail.com d79af41c43ea2a02afbbd66af85a70bed5b5757f Kris Marrier 24 24 20 18 2016-04-10 09:33:23.000 5.00 47.60 2016-04-13 09:34:06.000 24 Visa 4111111111111111 05/2017 24 19 minna_amigon@yahoo.c

The above code cell only returns 41 records, which makes sense since that is how many orders were placed, but we definitely aren't seeing all of our customers since our first set of queries showed 485 different customers. We can use a left join to pull those customers back in. The way that this works is that we are telling the query to return all records from the "left" side of the join statement, so we will leave customers on the left side for now.

In [19]:
--Use left join to show all customers as well as order information for customers that have ordered
SELECT c.*, o.*
FROM gs_customers as c
LEFT JOIN gs_orders as o ON c.customer_id = o.customer_id

(491 rows affected)

Total execution time: 00:00:00.340

customer_id email_address customer_password first_name last_name shipping_address_id billing_address_id order_id customer_id order_date ship_amount tax_amount ship_date ship_address_id card_type card_number card_expires billing_address_id 1 allan.sherwood@yahoo.com c44321e51ec184a2f739318639cec426de774451 Allan Sherwood 1 2 3 1 2016-03-29 09:44:58.000 10.00 102.29 2016-04-01 09:45:41.000 1 Visa 4111111111111111 06/2020 2 1 allan.sherwood@yahoo.com c44321e51ec184a2f739318639cec426de774451 Allan Sherwood 1 2 1 1 2016-03-28 09:40:28.000 5.00 58.75 2016-03-31 09:41:11.000 1 Visa 4111111111111111 04/2018 2 2 barryz@gmail.com d9e03c0b34c57d034edda004ec8bae5d53667e36 Barry Zimmer 3 3 2 2 2016-03-28 11:23:20.000 5.00 21.27 2016-03-31 11:24:03.000 3 Visa 4012888888881881 08/2020 3 3 christineb@solarone.com 13ef4f968693bda97a898ece497da087b182808e Christine Brown 4 4 4 3 2016-03-30 15:22:31.000 10.00 117.50 2016-04-02 15:23:14.000 4 American Express 3782822463100005 02/2017 4 4 david.goldstein@hotmail.com 2a367cbb171d78d293f40fd7d1defb31e3fb1728 David Goldstein 5 6 5 4 2016-03-31 05:43:11.000 5.00 20.93 2016-04-03 05:43:54.000 5 Visa 4111111111111111 09/2019 6 4 david.goldstein@hotmail.com 2a367cbb171d78d293f40fd7d1defb31e3fb1728 David Goldstein 5 6 9 4 2016-04-03 12:22:31.000 15.00 102.75 2016-04-06 12:23:14.000 5 Visa 4111111111111111 01/2020 6 4 david.goldstein@hotmail.com 2a367cbb171d78d293f40fd7d1defb31e3fb1728 David Goldstein 5 6 29 4 2016-04-25 23:36:41.000 25.00 196.00 2016-04-28 23:37:24.000 5 Visa 4012888888881881 03/2017 6 5 erinv@gmail.com 2e203dd22e39e3a8930e7641fe074fec2b18b102 Erin Valentino 7 7 6 5 2016-03-31 18:37:22.000 5.00 20.93 2016-04-03 18:38:05.000 7 Discover 6011111111111117 04/2020 7 6 frankwilson@sbcglobal.net b13773cfee62f832cacb618b257feec972f30b13 Frank Lee Wilson 8 8 7 6 2016-04-01 23:11:12.000 15.00 107.80 2016-04-04 23:11:55.000 8 MasterCard 5555555555554444 12/2018 8 7 gary_hernandez@yahoo.com e931eea39d638c0324c0065c40e2c0acc91ceca9 Gary Hernandez 9 10 8 7 2016-04-02 11:26:38.000 5.00 47.60 2016-04-05 11:27:21.000 9 Visa 4012888888881881 04/2017 10 8 heatheresway@mac.com 1867b31afdfbb6814133aa545f67305bc2f211b6 Heather Esway 11 12 22 8 2016-04-12 12:26:52.000 5.00 84.57 2016-04-15 12:27:35.000 11 Visa 4111111111111111 03/2020 12 8 heatheresway@mac.com 1867b31afdfbb6814133aa545f67305bc2f211b6 Heather Esway 11 12 10 8 2016-04-03 14:59:20.000 5.00 26.25 2016-04-06 15:00:03.000 11 Visa 4111111111111111 08/2019 12 9 jbutt@gmail.com 28592e5b54a8babc3cb6ad0a1c9094a2621c8ac3 James Butt 13 13 11 9 2016-04-04 06:24:44.000 5.00 34.25 2016-04-07 06:25:27.000 13 Visa 4012888888881881 08/2019 13 10 josephine_darakjy@darakjy.org 815d965d07c98821d8ca725243bd09def4e33f24 Josephine Darakjy 14 14 12 10 2016-04-04 08:15:12.000 5.00 84.57 2016-04-07 08:15:55.000 14 Visa 4111111111111111 03/2017 14 11 art@venere.org 39f7cfd0b51970b6bcd55a00da8e672b7153a183 Art Venere 15 16 13 11 2016-04-04 11:20:31.000 5.00 47.60 2016-04-07 11:21:14.000 15 Visa 4111111111111111 02/2020 16 12 lpaprocki@hotmail.com af27e4c5b48549d6dee6199e56834af1404f6169 Lenna Paprocki 17 17 14 12 2016-04-05 09:24:53.000 10.00 117.50 2016-04-08 09:25:36.000 17 Visa 4111111111111111 11/2018 17 13 donette.foller@cox.net 356584acaa7164280c97b147f317ce0d76d5993c Donette Foller 18 18 15 13 2016-04-05 14:52:17.000 5.00 39.20 2016-04-08 14:53:00.000 18 American Express 3782822463100005 02/2018 18 14 simona@morasca.com 81133e972ea9b42831ef1fe7a846493526bfd3c3 Simona Morasca 19 19 16 14 2016-04-06 07:53:42.000 10.00 51.97 2016-04-09 07:54:25.000 19 Visa 4111111111111111 01/2019 19 15 mitsue_tollner@yahoo.com 460420ee0d4f211f6be0bf7e7f352cf135cac579 Mitsue Tollner 20 21 17 15 2016-04-06 17:24:28.000 5.00 34.25 2016-04-09 17:25:11.000 20 Visa 4111111111111111 07/2020 21 16 leota@hotmail.com ceb92dd2b07d22fa39b9188306b9b04a78d80fe8 Leota Dilliard 22 22 18 16 2016-04-06 18:41:53.000 5.00 34.25 2016-04-09 18:42:36.000 22 MasterCard 5555555555554444 12/2017 22 17 sage_wies

We can see that the query is now showing information for all customers regardless of if they placed an order or not, and there are a few duplicates for customers who placed multiple orders. Please scroll down to customer 37, and then scroll over to view the order data. Notice that all of their order data is Null because they never placed an order. Since there were no orders placed for these customers, there is nothing for the orders table to return here.

What would happen if we made this a right join instead?

In [20]:
--Use a right join to show all orders as well as customer information if there is a match
SELECT c.*, o.*
FROM gs_customers as c
RIGHT JOIN gs_orders as o ON c.customer_id = o.customer_id

(41 rows affected)

Total execution time: 00:00:00.312

customer_id email_address customer_password first_name last_name shipping_address_id billing_address_id order_id customer_id order_date ship_amount tax_amount ship_date ship_address_id card_type card_number card_expires billing_address_id 1 allan.sherwood@yahoo.com c44321e51ec184a2f739318639cec426de774451 Allan Sherwood 1 2 1 1 2016-03-28 09:40:28.000 5.00 58.75 2016-03-31 09:41:11.000 1 Visa 4111111111111111 04/2018 2 2 barryz@gmail.com d9e03c0b34c57d034edda004ec8bae5d53667e36 Barry Zimmer 3 3 2 2 2016-03-28 11:23:20.000 5.00 21.27 2016-03-31 11:24:03.000 3 Visa 4012888888881881 08/2020 3 1 allan.sherwood@yahoo.com c44321e51ec184a2f739318639cec426de774451 Allan Sherwood 1 2 3 1 2016-03-29 09:44:58.000 10.00 102.29 2016-04-01 09:45:41.000 1 Visa 4111111111111111 06/2020 2 3 christineb@solarone.com 13ef4f968693bda97a898ece497da087b182808e Christine Brown 4 4 4 3 2016-03-30 15:22:31.000 10.00 117.50 2016-04-02 15:23:14.000 4 American Express 3782822463100005 02/2017 4 4 david.goldstein@hotmail.com 2a367cbb171d78d293f40fd7d1defb31e3fb1728 David Goldstein 5 6 5 4 2016-03-31 05:43:11.000 5.00 20.93 2016-04-03 05:43:54.000 5 Visa 4111111111111111 09/2019 6 5 erinv@gmail.com 2e203dd22e39e3a8930e7641fe074fec2b18b102 Erin Valentino 7 7 6 5 2016-03-31 18:37:22.000 5.00 20.93 2016-04-03 18:38:05.000 7 Discover 6011111111111117 04/2020 7 6 frankwilson@sbcglobal.net b13773cfee62f832cacb618b257feec972f30b13 Frank Lee Wilson 8 8 7 6 2016-04-01 23:11:12.000 15.00 107.80 2016-04-04 23:11:55.000 8 MasterCard 5555555555554444 12/2018 8 7 gary_hernandez@yahoo.com e931eea39d638c0324c0065c40e2c0acc91ceca9 Gary Hernandez 9 10 8 7 2016-04-02 11:26:38.000 5.00 47.60 2016-04-05 11:27:21.000 9 Visa 4012888888881881 04/2017 10 4 david.goldstein@hotmail.com 2a367cbb171d78d293f40fd7d1defb31e3fb1728 David Goldstein 5 6 9 4 2016-04-03 12:22:31.000 15.00 102.75 2016-04-06 12:23:14.000 5 Visa 4111111111111111 01/2020 6 8 heatheresway@mac.com 1867b31afdfbb6814133aa545f67305bc2f211b6 Heather Esway 11 12 10 8 2016-04-03 14:59:20.000 5.00 26.25 2016-04-06 15:00:03.000 11 Visa 4111111111111111 08/2019 12 9 jbutt@gmail.com 28592e5b54a8babc3cb6ad0a1c9094a2621c8ac3 James Butt 13 13 11 9 2016-04-04 06:24:44.000 5.00 34.25 2016-04-07 06:25:27.000 13 Visa 4012888888881881 08/2019 13 10 josephine_darakjy@darakjy.org 815d965d07c98821d8ca725243bd09def4e33f24 Josephine Darakjy 14 14 12 10 2016-04-04 08:15:12.000 5.00 84.57 2016-04-07 08:15:55.000 14 Visa 4111111111111111 03/2017 14 11 art@venere.org 39f7cfd0b51970b6bcd55a00da8e672b7153a183 Art Venere 15 16 13 11 2016-04-04 11:20:31.000 5.00 47.60 2016-04-07 11:21:14.000 15 Visa 4111111111111111 02/2020 16 12 lpaprocki@hotmail.com af27e4c5b48549d6dee6199e56834af1404f6169 Lenna Paprocki 17 17 14 12 2016-04-05 09:24:53.000 10.00 117.50 2016-04-08 09:25:36.000 17 Visa 4111111111111111 11/2018 17 13 donette.foller@cox.net 356584acaa7164280c97b147f317ce0d76d5993c Donette Foller 18 18 15 13 2016-04-05 14:52:17.000 5.00 39.20 2016-04-08 14:53:00.000 18 American Express 3782822463100005 02/2018 18 14 simona@morasca.com 81133e972ea9b42831ef1fe7a846493526bfd3c3 Simona Morasca 19 19 16 14 2016-04-06 07:53:42.000 10.00 51.97 2016-04-09 07:54:25.000 19 Visa 4111111111111111 01/2019 19 15 mitsue_tollner@yahoo.com 460420ee0d4f211f6be0bf7e7f352cf135cac579 Mitsue Tollner 20 21 17 15 2016-04-06 17:24:28.000 5.00 34.25 2016-04-09 17:25:11.000 20 Visa 4111111111111111 07/2020 21 16 leota@hotmail.com ceb92dd2b07d22fa39b9188306b9b04a78d80fe8 Leota Dilliard 22 22 18 16 2016-04-06 18:41:53.000 5.00 34.25 2016-04-09 18:42:36.000 22 MasterCard 5555555555554444 12/2017 22 17 sage_wieser@cox.net 0c317d5469cda724d8b81ccd989179c99c3f2563 Sage Wieser 23 23 19 17 2016-04-08 12:21:31.000 10.00 117.50 2016-04-11 12:22:14.000 23 Visa 4012888888881881 12/2017 23 18 kris@gmail.com d79af41c43ea2a02afbbd66af85a70bed5b5757f Kris Marrier 24 24 20 18 2016-04-10 09:33:23.000 5.00 47.60 2016-04-13 09:34:06.000 24 Visa 4111111111111111 05/2017 24 19 minna_amigon@yahoo.c

We are back to seeing 41 records and we don't see as many Null values. This is because we are now showing all records on the right side of the join statement (orders) and only those on the left side that have a related record.

Putting together everything that we have learned, we now know enough to create a query that shows which customers haven't placed an order.

- Show all customers- Left join between customers and orders
- Only show customers who don't have an order- set criteria to Null for order data

Let's incoroporate this into our next query.

In [21]:
--Use left join to show only customers who have never placed an order
SELECT c.*, o.*
FROM gs_customers as c
LEFT JOIN gs_orders as o ON c.customer_id = o.customer_id
WHERE o.customer_id is NULL

(450 rows affected)

Total execution time: 00:00:00.308

customer_id email_address customer_password first_name last_name shipping_address_id billing_address_id order_id customer_id order_date ship_amount tax_amount ship_date ship_address_id card_type card_number card_expires billing_address_id 36 ezekiel@chui.com 28bb43c8489d35ae5e0410a3e7f2920184f86d79 Ezekiel Chui 46 46 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 37 wkusko@yahoo.com 3769a748f05734c0692cd7dcb1718b39dc0fdc52 Willow Kusko 47 47 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 38 bfigeroa@aol.com b016c40c80f68317b242d07469d65c6ba07310dd Bernardo Figeroa 48 48 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 39 ammie@corrio.com 88f982f0ad4f45293d4adb30294fe7407cf99c01 Ammie Corrio 49 49 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 40 francine_vocelka@vocelka.com 94e3fe0c1a2a02d28c64da77b339f14308f2ee29 Francine Vocelka 50 51 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 41 ernie_stenseth@aol.com 4944f2cfff1dc49827539f415a390c56d6185fbd Ernie Stenseth 52 52 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 42 albina@glick.com afc12cca2e0eb72a0d8ca7849ac2b8687fef3e63 Albina Glick 53 53 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 43 asergi@gmail.com 8b06eecc3b3ac8aef4199354845a340b5776b0fc Alishia Sergi 54 54 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 44 solange@shinko.com b067b71aff361ba9b6d1eadd27e52eebdcde26ca Solange Shinko 55 55 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 45 jose@yahoo.com 866b8c2845010932a0e954c8596d52791c9dd83d Jose Stockham 56 56 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 46 rozella.ostrosky@ostrosky.com 0c6e3b007e144cd464786b81a5e3b51d1eb9b1be Rozella Ostrosky 57 57 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 47 valentine_gillian@gmail.com 906c6f5df090604345b51cfaaa18be4318fd7f58 Valentine Gillian 58 59 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 48 kati.rulapaugh@hotmail.com 82a759a36c991fc2d0161403b79f623c2c4aee73 Kati Rulapaugh 60 60 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 49 youlanda@aol.com 65240df71caf6471f67eedbbf8137d391af9bde3 Youlanda Schemmer 61 61 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 50 doldroyd@aol.com 0875cab0a2ff9851313147fce35a52a3ac16b60e Dyan Oldroyd 62 62 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 51 roxane@hotmail.com ef097d882a433ac306f4d1ef90e9733ec8284863 Roxane Campain 63 63 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 52 lperin@perin.org 57944d769b84f9f74eda6b8056830cb175860ed6 Lavera Perin 64 64 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 53 erick.ferencz@aol.com 0d27e2fb7bc3dcfb3b395145a112d95decada35b Erick Ferencz 65 66 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 54 fsaylors@saylors.org 4d23046b89b9fd957b9e7798709d2f006e758fb0 Fatima Saylors 67 67 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 55 jina_briddick@briddick.com eaee27e032471a3748c5d3140864a79c37d60ede Jina Briddick 68 69 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 56 kanisha_waycott@yahoo.com 3724279c88980023b3b2a9350eabd86e57bb4054 Kanisha Waycott 70 70 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 57 emerson.bowley@bowley.org b7063bbdd35fd8ee545b787d1595aad80ce82617 Emerson Bowley 71 71 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 58 bmalet@yahoo.com 7a6cc6177bf062818821854db22d146dc90d27b1 Blair Malet 72 72 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 59 bbolognia@yahoo.com 0558a2848e0f5b1d9d5cfcfc925a9b550945036b Brock Bolognia 73 73 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 60 lnestle@hotmail.com b04f6a00ae7dbf47a1aeeda97cd7fe662da04513 Lorrie Nestle 74 74 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 61 sabra@uyetake.org d373887fa37c7d724a92913ffd0e977e98d18acb Sabra Uyetake 75 75 NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL NULL 62 mmastella@mastella.com 83df6846b4c6c7a5de0b39cd681fa508236a7979 Marjory Mastella 76 76 NULL NULL NULL NULL

By adding a where clause to only show Nulls from Orders.CustomerID we were able to filter out any customers that had related data in the orders table, or in other words, had placed an order. Now we know that there are 450 customers in our database that have yet to place an order.

✏️ **Practice**

Edit the below query to show all customers who do not have an address in the addresses table.

In [24]:
SELECT c.*
FROM gs_customers AS c
LEFT JOIN gs_addresses AS a 
    ON c.customer_id = a.customer_id
WHERE a.customer_id IS NULL;

(1 row affected)

Total execution time: 00:00:00.270

customer_id email_address customer_password first_name last_name shipping_address_id billing_address_id 325 badkin@hotmail.com 6b1b092e1a4f9b67b13b92ffde100ba21bc5f88e Barbra Adkin 349 349

## **Junction/Join Tables**

Junciton, or join, tables are a special type of table that helps us resolve a many-to-many relationship. In the _guitar store_ example we have a many-to-many relationship between products and orders. This is because an individual product can be part of many orders, and an order can be made of many products. For example, Order #31 may contain products 1 and 6. At first this looks like a one-to-many relationship, however, Order # 41 contains products 6 and 8. We have multiple products appearing in different orders, which makes this a many-to-many relationship. To resolve this, we use a junction table. This junctino table, _OrderItems_, lists the OrderID and each associated ProductID in their own row so that we can query them. If we run the below code cells (2) we can see that there are no fields that we can use to join Prducts and Orders. They have no common data.

In [25]:
SELECT *
FROM gs_products AS p

(10 rows affected)

Total execution time: 00:00:01.141

product_id category_id product_code product_name product_description list_price discount_percent date_added 1 1 strat Fender Stratocaster The Fender Stratocaster is the electric guitar design that changed the world. New features include a tinted neck, parchment pickguard and control knobs, and a '70s-style logo. Includes select alder body, 21-fret maple neck with your choice of a rosewood or maple fretboard, 3 single-coil pickups, vintage-style tremolo, and die-cast tuning keys. This guitar features a thicker bridge block for increased sustain and a more stable point of contact with the strings. At this low price, why play anything but the real thing?\r\n\r\nFeatures:\r\n\r\n* New features:\r\n* Thicker bridge block\r\n* 3-ply parchment pick guard\r\n* Tinted neck 699.00 30.00 2015-10-30 09:32:40.000 2 1 les_paul Gibson Les Paul This Les Paul guitar offers a carved top and humbucking pickups. It has a simple yet elegant design. Cutting-yet-rich tone?the hallmark of the Les Paul?pours out of the 490R and 498T Alnico II magnet humbucker pickups, which are mounted on a carved maple top with a mahogany back. The faded finish models are equipped with BurstBucker Pro pickups and a mahogany top. This guitar includes a Gibson hardshell case (Faded and satin finish models come with a gig bag) and a limited lifetime warranty.\r\n\r\nFeatures:\r\n\r\n* Carved maple top and mahogany back (Mahogany top on faded finish models)\r\n* Mahogany neck, '59 Rounded Les Paul\r\n* Rosewood fingerboard (Ebony on Alpine white)\r\n* Tune-O-Matic bridge with stopbar\r\n* Chrome or gold hardware\r\n* 490R and 498T Alnico 2 magnet humbucker pickups (BurstBucker Pro on faded finish models)\r\n* 2 volume and 2 tone knobs, 3-way switch 1199.00 30.00 2015-12-05 16:33:13.000 3 1 sg Gibson SG This Gibson SG electric guitar takes the best of the '62 original and adds the longer and sturdier neck joint of the late '60s models. All the classic features you'd expect from a historic guitar. Hot humbuckers go from rich, sweet lightning to warm, tingling waves of sustain. A silky-fast rosewood fretboard plays like a dream. The original-style beveled mahogany body looks like a million bucks. Plus, Tune-O-Matic bridge and chrome hardware. Limited lifetime warranty. Includes hardshell case.\r\n\r\nFeatures:\r\n\r\n* Double-cutaway beveled mahogany body\r\n* Set mahogany neck with rounded '50s profile\r\n* Bound rosewood fingerboard with trapezoid inlays\r\n* Tune-O-Matic bridge with stopbar tailpiece\r\n* Chrome hardware\r\n* 490R humbucker in the neck position\r\n* 498T humbucker in the bridge position\r\n* 2 volume knobs, 2 tone knobs, 3-way switch\r\n* 24-3/4" scale 2517.00 52.00 2016-02-04 11:04:31.000 4 1 fg700s Yamaha FG700S The Yamaha FG700S solid top acoustic guitar has the ultimate combo for projection and pure tone. The expertly braced spruce top speaks clearly atop the rosewood body. It has a rosewood fingerboard, rosewood bridge, die-cast tuners, body and neck binding, and a tortoise pickguard.\r\n\r\nFeatures:\r\n\r\n* Solid Sitka spruce top\r\n* Rosewood back and sides\r\n* Rosewood fingerboard\r\n* Rosewood bridge\r\n* White/black body and neck binding\r\n* Die-cast tuners\r\n* Tortoise pickguard\r\n* Limited lifetime warranty 489.99 38.00 2016-06-01 11:12:59.000 5 1 washburn Washburn D10S The Washburn D10S acoustic guitar is superbly crafted with a solid spruce top and mahogany back and sides for exceptional tone. A mahogany neck and rosewood fingerboard make fretwork a breeze, while chrome Grover-style machines keep you perfectly tuned. The Washburn D10S comes with a limited lifetime warranty.\r\n\r\nFeatures:\r\n\r\n * Spruce top\r\n * Mahogany back, sides\r\n * Mahogany neck Rosewood fingerboard\r\n * Chrome Grover-style machines 299.00 0.00 2016-07-30 13:58:35.000 6 1 rodriguez Rodriguez Caballero 11 Featuring a carefully chosen, solid Canadian cedar top and laminated bubinga back and sides, the Caballero 11 classical guitar is a beauty to behold and 

In [26]:
SELECT *
FROM gs_orders as o

(41 rows affected)

Total execution time: 00:00:00.292

order_id customer_id order_date ship_amount tax_amount ship_date ship_address_id card_type card_number card_expires billing_address_id 1 1 2016-03-28 09:40:28.000 5.00 58.75 2016-03-31 09:41:11.000 1 Visa 4111111111111111 04/2018 2 2 2 2016-03-28 11:23:20.000 5.00 21.27 2016-03-31 11:24:03.000 3 Visa 4012888888881881 08/2020 3 3 1 2016-03-29 09:44:58.000 10.00 102.29 2016-04-01 09:45:41.000 1 Visa 4111111111111111 06/2020 2 4 3 2016-03-30 15:22:31.000 10.00 117.50 2016-04-02 15:23:14.000 4 American Express 3782822463100005 02/2017 4 5 4 2016-03-31 05:43:11.000 5.00 20.93 2016-04-03 05:43:54.000 5 Visa 4111111111111111 09/2019 6 6 5 2016-03-31 18:37:22.000 5.00 20.93 2016-04-03 18:38:05.000 7 Discover 6011111111111117 04/2020 7 7 6 2016-04-01 23:11:12.000 15.00 107.80 2016-04-04 23:11:55.000 8 MasterCard 5555555555554444 12/2018 8 8 7 2016-04-02 11:26:38.000 5.00 47.60 2016-04-05 11:27:21.000 9 Visa 4012888888881881 04/2017 10 9 4 2016-04-03 12:22:31.000 15.00 102.75 2016-04-06 12:23:14.000 5 Visa 4111111111111111 01/2020 6 10 8 2016-04-03 14:59:20.000 5.00 26.25 2016-04-06 15:00:03.000 11 Visa 4111111111111111 08/2019 12 11 9 2016-04-04 06:24:44.000 5.00 34.25 2016-04-07 06:25:27.000 13 Visa 4012888888881881 08/2019 13 12 10 2016-04-04 08:15:12.000 5.00 84.57 2016-04-07 08:15:55.000 14 Visa 4111111111111111 03/2017 14 13 11 2016-04-04 11:20:31.000 5.00 47.60 2016-04-07 11:21:14.000 15 Visa 4111111111111111 02/2020 16 14 12 2016-04-05 09:24:53.000 10.00 117.50 2016-04-08 09:25:36.000 17 Visa 4111111111111111 11/2018 17 15 13 2016-04-05 14:52:17.000 5.00 39.20 2016-04-08 14:53:00.000 18 American Express 3782822463100005 02/2018 18 16 14 2016-04-06 07:53:42.000 10.00 51.97 2016-04-09 07:54:25.000 19 Visa 4111111111111111 01/2019 19 17 15 2016-04-06 17:24:28.000 5.00 34.25 2016-04-09 17:25:11.000 20 Visa 4111111111111111 07/2020 21 18 16 2016-04-06 18:41:53.000 5.00 34.25 2016-04-09 18:42:36.000 22 MasterCard 5555555555554444 12/2017 22 19 17 2016-04-08 12:21:31.000 10.00 117.50 2016-04-11 12:22:14.000 23 Visa 4012888888881881 12/2017 23 20 18 2016-04-10 09:33:23.000 5.00 47.60 2016-04-13 09:34:06.000 24 Visa 4111111111111111 05/2017 24 21 19 2016-04-11 08:21:32.000 10.00 39.20 2016-04-14 08:22:15.000 25 Discover 6011111111111117 02/2019 26 22 8 2016-04-12 12:26:52.000 5.00 84.57 2016-04-15 12:27:35.000 11 Visa 4111111111111111 03/2020 12 23 20 2016-04-14 07:59:31.000 5.00 34.25 2016-04-17 08:00:14.000 27 Visa 4012888888881881 03/2019 27 24 21 2016-04-17 17:40:22.000 5.00 34.25 2016-04-20 17:41:05.000 28 Visa 4111111111111111 04/2018 28 25 22 2016-04-20 08:23:32.000 10.00 117.50 2016-04-23 08:24:15.000 29 Visa 4111111111111111 09/2018 29 26 23 2016-04-20 08:14:45.000 5.00 0.00 2016-04-23 08:15:28.000 30 American Express 3782822463100005 08/2017 30 27 24 2016-04-20 09:17:52.000 5.00 84.57 2016-04-23 09:18:35.000 31 Visa 4111111111111111 02/2018 31 28 25 2016-04-21 17:52:24.000 5.00 34.30 2016-04-24 17:53:07.000 32 Visa 4111111111111111 08/2019 32 29 4 2016-04-25 23:36:41.000 25.00 196.00 2016-04-28 23:37:24.000 5 Visa 4012888888881881 03/2017 6 30 26 2016-04-27 16:21:31.000 5.00 26.25 2016-04-30 16:22:14.000 33 Visa 4111111111111111 02/2018 33 31 27 2016-04-29 06:47:14.000 10.00 118.82 2016-05-02 06:47:57.000 34 Visa 4111111111111111 01/2018 34 32 18 2016-05-01 01:23:23.000 5.00 84.57 NULL 24 Discover 6011111111111117 02/2018 24 33 28 2016-05-01 09:11:51.000 10.00 41.86 2016-05-04 09:12:34.000 35 American Express 3782822463100005 04/2017 35 34 29 2016-05-02 11:36:12.000 5.00 58.75 2016-05-05 11:36:55.000 36 Visa 4111111111111111 06/2017 37 35 30 2016-05-04 03:52:23.000 5.00 39.20 2016-05-07 03:53:06.000 38 Visa 4111111111111111 09/2018 38 36 31 2016-05-04 12:31:33.000 5.00 21.27 2016-05-07 12:32:16.000 39 Visa 4012888888881881 11/2018 39 37 32 2016-05-06 14:15:21.000 5.00 84.57 2016-05-09 14:16:04.000 40 MasterCard 5555555555554444 02/2017 41 38 33 2016-05-08 11:41:24.000 10.00 117.50 NULL 42 Visa 4111111111111111 04/2018 43 39 29 

Now, if we look at the _OrderItems_ table then we can see that it contains information from both tables. This is showing which items (item\_id) were part of which orders (order\_id)

In [27]:
SELECT *
FROM gs_order_items AS oi

(47 rows affected)

Total execution time: 00:00:00.338

item_id order_id product_id item_price discount_amount quantity 1 1 2 1199.00 359.70 1 2 2 8 489.99 186.20 1 3 3 1 2517.00 1308.84 1 4 3 9 415.00 161.85 1 5 4 2 1199.00 359.70 2 6 5 10 299.00 0.00 1 7 6 10 299.00 0.00 1 8 7 5 699.99 210.00 1 9 7 3 799.99 240.00 1 10 7 5 699.99 210.00 1 11 8 4 799.99 120.00 1 12 9 6 699.00 209.70 3 13 10 7 499.99 125.00 1 14 11 6 699.00 209.70 1 15 12 1 2517.00 1308.84 1 16 13 4 799.99 120.00 1 17 14 2 1199.00 359.70 2 18 15 3 799.99 240.00 1 19 16 6 699.00 209.70 1 20 16 9 415.00 161.85 1 21 17 6 699.00 209.70 1 22 18 6 699.00 209.70 1 23 19 2 1199.00 359.70 2 24 20 4 799.99 120.00 1 25 21 3 799.99 240.00 1 26 22 1 2517.00 1308.84 1 27 23 6 699.00 209.70 1 28 24 6 699.00 209.70 1 29 25 2 1199.00 359.70 2 30 26 6 699.00 209.70 1 31 27 1 2517.00 1308.84 1 32 28 5 699.99 210.00 1 33 29 3 799.99 240.00 5 34 30 7 499.99 125.00 1 35 31 1 2517.00 1308.84 1 36 31 6 699.00 209.70 1 37 32 1 2517.00 1308.84 1 38 33 10 299.00 0.00 2 39 34 2 1199.00 359.70 1 40 35 3 799.99 240.00 1 41 36 8 489.99 186.20 1 42 37 1 2517.00 1308.84 1 43 38 2 1199.00 359.70 2 44 39 6 699.00 209.70 1 45 40 6 699.00 209.70 1 46 41 8 489.99 186.20 1 47 41 6 699.00 209.70 1

This table also includes useful information such as what the discount on the product was at the time of the order, the price that was paid, and the quantity purchased. We can think of this as a transaction table that is storing transaction level data for each product in each order. It is the most granular stage of our data.

If we want to connect products to orders, then we will need to join each of our products and orders tables to the OrderItems table. This can be done by using multiple join statements like in the cell below. At the end of one join statement, we just start the next one. The below code cell is showing the productID and orderID for each product/order pairing. If you scroll up, you will notice that these are the same combinations as what is in the ItemOrders table, but we are pulling from the Products and Orders table now instead!

In [28]:
SELECT p,product_id, o.order_id
FROM gs_products as p
JOIN gs_order_items as oi ON p.product_id = oi.product_id
JOIN gs_orders as o ON oi.order_id = o.order_id

: Msg 207, Level 16, State 1, Line 1
Invalid column name 'p'.

: Msg 209, Level 16, State 1, Line 1
Ambiguous column name 'product_id'.

Total execution time: 00:00:00.315

**✏️ Practice**

Edit the below code cell so that you are returning the name of each product, as well as the customerID of the customer who placed the order, and the OrderDate.

In [29]:
SELECT 
    p.product_name,
    o.customer_id,
    o.order_date
FROM gs_products AS p
JOIN gs_order_items AS oi 
    ON p.product_id = oi.product_id
JOIN gs_orders AS o
    ON oi.order_id = o.order_id;

(47 rows affected)

Total execution time: 00:00:00.297

product_name customer_id order_date Gibson Les Paul 1 2016-03-28 09:40:28.000 Hofner Icon 2 2016-03-28 11:23:20.000 Fender Stratocaster 1 2016-03-29 09:44:58.000 Ludwig 5-piece Drum Set with Cymbals 1 2016-03-29 09:44:58.000 Gibson Les Paul 3 2016-03-30 15:22:31.000 Tama 5-Piece Drum Set with Cymbals 4 2016-03-31 05:43:11.000 Tama 5-Piece Drum Set with Cymbals 5 2016-03-31 18:37:22.000 Washburn D10S 6 2016-04-01 23:11:12.000 Gibson SG 6 2016-04-01 23:11:12.000 Washburn D10S 6 2016-04-01 23:11:12.000 Yamaha FG700S 7 2016-04-02 11:26:38.000 Rodriguez Caballero 11 4 2016-04-03 12:22:31.000 Fender Precision 8 2016-04-03 14:59:20.000 Rodriguez Caballero 11 9 2016-04-04 06:24:44.000 Fender Stratocaster 10 2016-04-04 08:15:12.000 Yamaha FG700S 11 2016-04-04 11:20:31.000 Gibson Les Paul 12 2016-04-05 09:24:53.000 Gibson SG 13 2016-04-05 14:52:17.000 Rodriguez Caballero 11 14 2016-04-06 07:53:42.000 Ludwig 5-piece Drum Set with Cymbals 14 2016-04-06 07:53:42.000 Rodriguez Caballero 11 15 2016-04-06 17:24:28.000 Rodriguez Caballero 11 16 2016-04-06 18:41:53.000 Gibson Les Paul 17 2016-04-08 12:21:31.000 Yamaha FG700S 18 2016-04-10 09:33:23.000 Gibson SG 19 2016-04-11 08:21:32.000 Fender Stratocaster 8 2016-04-12 12:26:52.000 Rodriguez Caballero 11 20 2016-04-14 07:59:31.000 Rodriguez Caballero 11 21 2016-04-17 17:40:22.000 Gibson Les Paul 22 2016-04-20 08:23:32.000 Rodriguez Caballero 11 23 2016-04-20 08:14:45.000 Fender Stratocaster 24 2016-04-20 09:17:52.000 Washburn D10S 25 2016-04-21 17:52:24.000 Gibson SG 4 2016-04-25 23:36:41.000 Fender Precision 26 2016-04-27 16:21:31.000 Fender Stratocaster 27 2016-04-29 06:47:14.000 Rodriguez Caballero 11 27 2016-04-29 06:47:14.000 Fender Stratocaster 18 2016-05-01 01:23:23.000 Tama 5-Piece Drum Set with Cymbals 28 2016-05-01 09:11:51.000 Gibson Les Paul 29 2016-05-02 11:36:12.000 Gibson SG 30 2016-05-04 03:52:23.000 Hofner Icon 31 2016-05-04 12:31:33.000 Fender Stratocaster 32 2016-05-06 14:15:21.000 Gibson Les Paul 33 2016-05-08 11:41:24.000 Rodriguez Caballero 11 29 2016-05-08 22:22:26.000 Rodriguez Caballero 11 34 2016-05-08 21:41:29.000 Hofner Icon 35 2016-05-09 07:52:55.000 Rodriguez Caballero 11 35 2016-05-09 07:52:55.000

## **Exercises**

If you have not already, I would strongly recommend having access to the ERD Linked in Canvas for the Exercises.

1\. Using the below code cell, return the order date, ship amount, and ship date for every order, as well as the name of the customer that placed the order.

**First 5 rows of correct solution**
|order_date|ship_amount|ship_date|first_name|last_name|
|---|---|---|---|---|
|2016-03-28 09:40:28.000|5.00|2016-03-31 09:41:11.000|Allan|Sherwood|
|2016-03-28 11:23:20.000|5.00|2016-03-31 11:24:03.000|Barry|Zimmer|
|2016-03-29 09:44:58.000|10.00|2016-04-01 09:45:41.000|Allan|Sherwood|
|2016-03-30 15:22:31.000|10.00|2016-04-02 15:23:14.000|Christine|Brown|
|2016-03-31 05:43:11.000|5.00|2016-04-03 05:43:54.000|David|Goldstein|

In [30]:
--Insert your code below this line. You can make your own comments by using two hyphens
-- Return order date, ship amount, ship date, and customer name for each order
SELECT 
    o.order_date,
    o.ship_amount,
    o.ship_date,
    c.first_name,
    c.last_name
FROM gs_orders AS o
JOIN gs_customers AS c 
    ON o.customer_id = c.customer_id;

(41 rows affected)

Total execution time: 00:00:00.308

order_date ship_amount ship_date first_name last_name 2016-03-28 09:40:28.000 5.00 2016-03-31 09:41:11.000 Allan Sherwood 2016-03-28 11:23:20.000 5.00 2016-03-31 11:24:03.000 Barry Zimmer 2016-03-29 09:44:58.000 10.00 2016-04-01 09:45:41.000 Allan Sherwood 2016-03-30 15:22:31.000 10.00 2016-04-02 15:23:14.000 Christine Brown 2016-03-31 05:43:11.000 5.00 2016-04-03 05:43:54.000 David Goldstein 2016-03-31 18:37:22.000 5.00 2016-04-03 18:38:05.000 Erin Valentino 2016-04-01 23:11:12.000 15.00 2016-04-04 23:11:55.000 Frank Lee Wilson 2016-04-02 11:26:38.000 5.00 2016-04-05 11:27:21.000 Gary Hernandez 2016-04-03 12:22:31.000 15.00 2016-04-06 12:23:14.000 David Goldstein 2016-04-03 14:59:20.000 5.00 2016-04-06 15:00:03.000 Heather Esway 2016-04-04 06:24:44.000 5.00 2016-04-07 06:25:27.000 James Butt 2016-04-04 08:15:12.000 5.00 2016-04-07 08:15:55.000 Josephine Darakjy 2016-04-04 11:20:31.000 5.00 2016-04-07 11:21:14.000 Art Venere 2016-04-05 09:24:53.000 10.00 2016-04-08 09:25:36.000 Lenna Paprocki 2016-04-05 14:52:17.000 5.00 2016-04-08 14:53:00.000 Donette Foller 2016-04-06 07:53:42.000 10.00 2016-04-09 07:54:25.000 Simona Morasca 2016-04-06 17:24:28.000 5.00 2016-04-09 17:25:11.000 Mitsue Tollner 2016-04-06 18:41:53.000 5.00 2016-04-09 18:42:36.000 Leota Dilliard 2016-04-08 12:21:31.000 10.00 2016-04-11 12:22:14.000 Sage Wieser 2016-04-10 09:33:23.000 5.00 2016-04-13 09:34:06.000 Kris Marrier 2016-04-11 08:21:32.000 10.00 2016-04-14 08:22:15.000 Minna Amigon 2016-04-12 12:26:52.000 5.00 2016-04-15 12:27:35.000 Heather Esway 2016-04-14 07:59:31.000 5.00 2016-04-17 08:00:14.000 Abel Maclead 2016-04-17 17:40:22.000 5.00 2016-04-20 17:41:05.000 Kiley Caldarera 2016-04-20 08:23:32.000 10.00 2016-04-23 08:24:15.000 Graciela Ruta 2016-04-20 08:14:45.000 5.00 2016-04-23 08:15:28.000 Cammy Albares 2016-04-20 09:17:52.000 5.00 2016-04-23 09:18:35.000 Mattie Poquette 2016-04-21 17:52:24.000 5.00 2016-04-24 17:53:07.000 Meaghan Garufi 2016-04-25 23:36:41.000 25.00 2016-04-28 23:37:24.000 David Goldstein 2016-04-27 16:21:31.000 5.00 2016-04-30 16:22:14.000 Gladys Rim 2016-04-29 06:47:14.000 10.00 2016-05-02 06:47:57.000 Yuki Whobrey 2016-05-01 01:23:23.000 5.00 NULL Kris Marrier 2016-05-01 09:11:51.000 10.00 2016-05-04 09:12:34.000 Fletcher Flosi 2016-05-02 11:36:12.000 5.00 2016-05-05 11:36:55.000 Bette Nicka 2016-05-04 03:52:23.000 5.00 2016-05-07 03:53:06.000 Veronika Inouye 2016-05-04 12:31:33.000 5.00 2016-05-07 12:32:16.000 Willard Kolmetz 2016-05-06 14:15:21.000 5.00 2016-05-09 14:16:04.000 Maryann Royster 2016-05-08 11:41:24.000 10.00 NULL Alisha Slusarski 2016-05-08 22:22:26.000 5.00 NULL Bette Nicka 2016-05-08 21:41:29.000 5.00 NULL Allene Iturbide 2016-05-09 07:52:55.000 10.00 NULL Chanel Caudy

2\. Using the below code cell, return the first name, last name, and email address for each customer who has placed and order and has a yahoo email address.

**First 5 rows of correct solution**
|first_name|last_name|email_address|
|---|---|---|
|Allan|Sherwood|allan.sherwood@yahoo.com|
|Allan|Sherwood|allan.sherwood@yahoo.com|
|Gary|Hernandez|gary_hernandez@yahoo.com|
|Mitsue|Tollner|mitsue_tollner@yahoo.com|
|Minna|Amigon|minna_amigon@yahoo.com|

In [32]:
--Insert your code below this line. You can make your own comments by using two hyphens
-- Return first name, last name, and email for customers with yahoo email who have placed an order
SELECT 
    c.first_name,
    c.last_name,
    c.email_address
FROM gs_customers AS c
JOIN gs_orders AS o
    ON c.customer_id = o.customer_id
WHERE c.email_address LIKE '%@yahoo.com';

(6 rows affected)

Total execution time: 00:00:00.292

first_name last_name email_address Allan Sherwood allan.sherwood@yahoo.com Allan Sherwood allan.sherwood@yahoo.com Gary Hernandez gary_hernandez@yahoo.com Mitsue Tollner mitsue_tollner@yahoo.com Minna Amigon minna_amigon@yahoo.com Fletcher Flosi fletcher.flosi@yahoo.com

3\. Using the below code cell, return the order\_id and order date for items ordered by Kris Marrier

**First 5 rows of correct solution**

| order\_id | order\_date |
| --- | --- |
| 20 | 2016-04-10 09:33:23.000 |
| 32 | 2016-05-01 01:23:23.000 |

In [33]:
--Insert your code below this line. You can make your own comments by using two hyphens
SELECT 
    o.order_id,
    o.order_date
FROM gs_orders AS o
JOIN gs_customers AS c
    ON o.customer_id = c.customer_id
WHERE c.first_name = 'Kris'
  AND c.last_name = 'Marrier';

(2 rows affected)

Total execution time: 00:00:00.508

order_id order_date 20 2016-04-10 09:33:23.000 32 2016-05-01 01:23:23.000

4\. Expand the query from exercise 3 to also include the product name, price, and quantity. You will have to join a total of four tables.

**First 5 rows of correct solution**
|order_id|order_date|product_name|item_price|quantity|
|---|---|---|---|---|
|20|2016-04-10 09:33:23.000|Yamaha FG700S|799.99|1|
|32|2016-05-01 01:23:23.000|Fender Stratocaster|2517.00|1|


In [35]:
-- Return order_id, order_date, product name, item price, and quantity for orders placed by Kris Marrier
SELECT 
    o.order_id,
    o.order_date,
    p.product_name,
    oi.item_price,
    oi.quantity
FROM gs_orders AS o
JOIN gs_customers AS c
    ON o.customer_id = c.customer_id
JOIN gs_order_items AS oi
    ON o.order_id = oi.order_id
JOIN gs_products AS p
    ON oi.product_id = p.product_id
WHERE c.first_name = 'Kris'
  AND c.last_name = 'Marrier';

(2 rows affected)

Total execution time: 00:00:00.958

order_id order_date product_name item_price quantity 20 2016-04-10 09:33:23.000 Yamaha FG700S 799.99 1 32 2016-05-01 01:23:23.000 Fender Stratocaster 2517.00 1

## **Scenario**

The owner of Guitar Store wants you to create a query that shows all of the needed information for creating a batch of invoices. Your query should contain the Cateegory Name, Products Name, Description, List Price, Item Price, Discount Amount, Quantity, OrderDate, Shipping and tax costs, Customer name, Customer Shipping address, and their email.

Create your query in the code cell below:

**First 5 rows of correct solution**
|category_name|product_name|product_description|list_price|item_price|discount_amount|quantity|order_date|ship_amount|tax_amount|first_name|last_name|Line1|line2|city|state_code|zip_code|email_address|
|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
|Guitars|Gibson Les Paul|This Les Paul guitar offers a carved top and humbucking pickups. It has a simple yet elegant design. Cutting-yet-rich tone?the hallmark of the Les Paul?pours out of the 490R and 498T Alnico II magnet humbucker pickups, which are mounted on a carved maple top with a mahogany back. The faded finish models are equipped with BurstBucker Pro pickups and a mahogany top. This guitar includes a Gibson hardshell case (Faded and satin finish models come with a gig bag) and a limited lifetime warranty.\r\n\r\nFeatures:\r\n\r\n* Carved maple top and mahogany back (Mahogany top on faded finish models)\r\n* Mahogany neck, &#39;59 Rounded Les Paul\r\n* Rosewood fingerboard (Ebony on Alpine white)\r\n* Tune-O-Matic bridge with stopbar\r\n* Chrome or gold hardware\r\n* 490R and 498T Alnico 2 magnet humbucker pickups (BurstBucker Pro on faded finish models)\r\n* 2 volume and 2 tone knobs, 3-way switch|1199.00|1199.00|359.70|1|2016-03-28 09:40:28.000|5.00|58.75|Allan|Sherwood|100 East Ridgewood Ave.||Paramus|NJ|07652|allan.sherwood@yahoo.com|
|Basses|Hofner Icon|With authentic details inspired by the original, the Hofner Icon makes the legendary violin bass available to the rest of us. Don&#39;t get the idea that this a just a &quot;nowhere man&quot; look-alike. This quality instrument features a real spruce top and beautiful flamed maple back and sides. The semi-hollow body and set neck will give you the warm, round tone you expect from the violin bass.\r\n\r\nFeatures:\r\n\r\n* Authentic details inspired by the original\r\n* Spruce top\r\n* Flamed maple back and sides\r\n* Set neck\r\n* Rosewood fretboard\r\n* 30&quot; scale\r\n* 22 frets\r\n* Dot inlay|499.99|489.99|186.20|1|2016-03-28 11:23:20.000|5.00|21.27|Barry|Zimmer|16285 Wendell St.||Omaha|NE|68135|barryz@gmail.com|
|Guitars|Fender Stratocaster|The Fender Stratocaster is the electric guitar design that changed the world. New features include a tinted neck, parchment pickguard and control knobs, and a &#39;70s-style logo. Includes select alder body, 21-fret maple neck with your choice of a rosewood or maple fretboard, 3 single-coil pickups, vintage-style tremolo, and die-cast tuning keys. This guitar features a thicker bridge block for increased sustain and a more stable point of contact with the strings. At this low price, why play anything but the real thing?\r\n\r\nFeatures:\r\n\r\n* New features:\r\n* Thicker bridge block\r\n* 3-ply parchment pick guard\r\n* Tinted neck|699.00|2517.00|1308.84|1|2016-03-29 09:44:58.000|10.00|102.29|Allan|Sherwood|100 East Ridgewood Ave.||Paramus|NJ|07652|allan.sherwood@yahoo.com|
|Drums|Ludwig 5-piece Drum Set with Cymbals|This product includes a Ludwig 5-piece drum set and a Zildjian starter cymbal pack.\r\n\r\nWith the Ludwig drum set, you get famous Ludwig quality. This set features a bass drum, two toms, a floor tom, and a snare?each with a wrapped finish. Drum hardware includes LA214FP bass pedal, snare stand, cymbal stand, hi-hat stand, and a throne.\r\n\r\nWith the Zildjian cymbal pack, you get a 14&quot; crash, 18&quot; crash/ride, and a pair of 13&quot; hi-hats. Sound grooves and round hammer strikes in a simple circular pattern on the top surface of these cymbals magnify the basic sound of the distinctive alloy.\r\n\r\nFeatures:\r\n\r\n* Famous Ludwig quality\r\n* Wrapped finishes\r\n* 22&quot; x 16&quot; kick drum\r\n* 12&quot; x 10&quot; and 13&quot; x 11&quot; toms\r\n* 16&quot; x 16&quot; floor tom\r\n* 14&quot; x 6-1/2&quot; snare drum kick pedal\r\n* Snare stand\r\n* Straight cymbal stand hi-hat stand\r\n* FREE throne|699.99|415.00|161.85|1|2016-03-29 09:44:58.000|10.00|102.29|Allan|Sherwood|100 East Ridgewood Ave.||Paramus|NJ|07652|allan.sherwood@yahoo.com|
|Guitars|Gibson Les Paul|This Les Paul guitar offers a carved top and humbucking pickups. It has a simple yet elegant design. Cutting-yet-rich tone?the hallmark of the Les Paul?pours out of the 490R and 498T Alnico II magnet humbucker pickups, which are mounted on a carved maple top with a mahogany back. The faded finish models are equipped with BurstBucker Pro pickups and a mahogany top. This guitar includes a Gibson hardshell case (Faded and satin finish models come with a gig bag) and a limited lifetime warranty.\r\n\r\nFeatures:\r\n\r\n* Carved maple top and mahogany back (Mahogany top on faded finish models)\r\n* Mahogany neck, &#39;59 Rounded Les Paul\r\n* Rosewood fingerboard (Ebony on Alpine white)\r\n* Tune-O-Matic bridge with stopbar\r\n* Chrome or gold hardware\r\n* 490R and 498T Alnico 2 magnet humbucker pickups (BurstBucker Pro on faded finish models)\r\n* 2 volume and 2 tone knobs, 3-way switch|1199.00|1199.00|359.70|2|2016-03-30 15:22:31.000|10.00|117.50|Christine|Brown|19270 NW Cornell Rd.||Beaverton|OR|97006|christineb@solarone.com|


In [39]:

-- Return all invoice batch details
SELECT 
    cat.category_name,
    p.product_name,
    p.product_description,
    p.list_price,
    oi.item_price,
    oi.discount_amount,
    oi.quantity,
    o.order_date,
    o.ship_amount,
    o.tax_amount,
    c.first_name,
    c.last_name,
    a.line1,
    a.line2,
    a.city,
    a.state_code,
    a.zip_code,
    c.email_address
FROM gs_orders AS o
JOIN gs_customers AS c
    ON o.customer_id = c.customer_id
JOIN gs_addresses AS a
    ON c.customer_id = a.customer_id
JOIN gs_order_items AS oi
    ON o.order_id = oi.order_id
JOIN gs_products AS p
    ON oi.product_id = p.product_id
JOIN gs_categories AS cat
    ON p.category_id = cat.category_id;


(63 rows affected)

Total execution time: 00:00:00.327

category_name product_name product_description list_price item_price discount_amount quantity order_date ship_amount tax_amount first_name last_name line1 line2 city state_code zip_code email_address Guitars Gibson Les Paul This Les Paul guitar offers a carved top and humbucking pickups. It has a simple yet elegant design. Cutting-yet-rich tone?the hallmark of the Les Paul?pours out of the 490R and 498T Alnico II magnet humbucker pickups, which are mounted on a carved maple top with a mahogany back. The faded finish models are equipped with BurstBucker Pro pickups and a mahogany top. This guitar includes a Gibson hardshell case (Faded and satin finish models come with a gig bag) and a limited lifetime warranty.\r\n\r\nFeatures:\r\n\r\n* Carved maple top and mahogany back (Mahogany top on faded finish models)\r\n* Mahogany neck, '59 Rounded Les Paul\r\n* Rosewood fingerboard (Ebony on Alpine white)\r\n* Tune-O-Matic bridge with stopbar\r\n* Chrome or gold hardware\r\n* 490R and 498T Alnico 2 magnet humbucker pickups (BurstBucker Pro on faded finish models)\r\n* 2 volume and 2 tone knobs, 3-way switch 1199.00 1199.00 359.70 1 2016-03-28 09:40:28.000 5.00 58.75 Allan Sherwood 100 East Ridgewood Ave. Paramus NJ 07652 allan.sherwood@yahoo.com Guitars Fender Stratocaster The Fender Stratocaster is the electric guitar design that changed the world. New features include a tinted neck, parchment pickguard and control knobs, and a '70s-style logo. Includes select alder body, 21-fret maple neck with your choice of a rosewood or maple fretboard, 3 single-coil pickups, vintage-style tremolo, and die-cast tuning keys. This guitar features a thicker bridge block for increased sustain and a more stable point of contact with the strings. At this low price, why play anything but the real thing?\r\n\r\nFeatures:\r\n\r\n* New features:\r\n* Thicker bridge block\r\n* 3-ply parchment pick guard\r\n* Tinted neck 699.00 2517.00 1308.84 1 2016-03-29 09:44:58.000 10.00 102.29 Allan Sherwood 100 East Ridgewood Ave. Paramus NJ 07652 allan.sherwood@yahoo.com Drums Ludwig 5-piece Drum Set with Cymbals This product includes a Ludwig 5-piece drum set and a Zildjian starter cymbal pack.\r\n\r\nWith the Ludwig drum set, you get famous Ludwig quality. This set features a bass drum, two toms, a floor tom, and a snare?each with a wrapped finish. Drum hardware includes LA214FP bass pedal, snare stand, cymbal stand, hi-hat stand, and a throne.\r\n\r\nWith the Zildjian cymbal pack, you get a 14" crash, 18" crash/ride, and a pair of 13" hi-hats. Sound grooves and round hammer strikes in a simple circular pattern on the top surface of these cymbals magnify the basic sound of the distinctive alloy.\r\n\r\nFeatures:\r\n\r\n* Famous Ludwig quality\r\n* Wrapped finishes\r\n* 22" x 16" kick drum\r\n* 12" x 10" and 13" x 11" toms\r\n* 16" x 16" floor tom\r\n* 14" x 6-1/2" snare drum kick pedal\r\n* Snare stand\r\n* Straight cymbal stand hi-hat stand\r\n* FREE throne 699.99 415.00 161.85 1 2016-03-29 09:44:58.000 10.00 102.29 Allan Sherwood 100 East Ridgewood Ave. Paramus NJ 07652 allan.sherwood@yahoo.com Guitars Gibson Les Paul This Les Paul guitar offers a carved top and humbucking pickups. It has a simple yet elegant design. Cutting-yet-rich tone?the hallmark of the Les Paul?pours out of the 490R and 498T Alnico II magnet humbucker pickups, which are mounted on a carved maple top with a mahogany back. The faded finish models are equipped with BurstBucker Pro pickups and a mahogany top. This guitar includes a Gibson hardshell case (Faded and satin finish models come with a gig bag) and a limited lifetime warranty.\r\n\r\nFeatures:\r\n\r\n* Carved maple top and mahogany back (Mahogany top on faded finish models)\r\n* Mahogany neck, '59 Rounded Les Paul\r\n* Rosewood fingerboard (Ebony on Alpine white)\r\n* Tune-O-Matic bridge with stopbar\r\n* Chrome or gold hardware\r\n* 490R and 498T Alnico 2 magnet humbucker pickups (BurstBucker Pro on faded finish models)\r\n* 2 volume 